File: framework/languages/java/e2e/SpotService/Client/src/main/java/systems/zlink/e2e/spotservice/client/Scenarios/SmD6Scenario.java
package systems.zlink.e2e.spotservice.client.Scenarios;

import java.time.Duration;
import java.util.List;
import java.util.UUID;
import systems.zlink.e2e.spotservice.shared.Contracts;
import systems.zlink.stream.connector.ZLinkStreamConnector;

public final class SmD6Scenario extends SpotServiceScenarioContext {
    private SmD6Scenario(SpotServiceScenarioContext context) {
        super(context);
    }

    public static void run(SpotServiceScenarioContext context) {
        new SmD6Scenario(context).execute();
    }

    private void execute() {
        String boundActorId = "actor-sm-d6-" + UUID.randomUUID().toString().replace("-", "");
        String shadowActorId = "actor-sm-d6-shadow-" + UUID.randomUUID().toString().replace("-", "");
        ZLinkStreamConnector bound = createStreamConnector(options().streamAEndpoint());
        ZLinkStreamConnector shadow = createStreamConnector(options().streamBEndpoint());
        Contracts.ActorProfile boundProfile = new Contracts.ActorProfile("Bound", 6, List.of("bound"));
        Contracts.ActorProfile shadowProfile = new Contracts.ActorProfile("Shadow", 6, List.of("shadow"));
        try {
            bound.connect().submit().toCompletableFuture().join();
            Contracts.ActorAuthRes boundAuth = bound
                .request(new Contracts.ActorAuthReq(boundActorId, boundProfile))
                .submit(Contracts.ActorAuthRes.class).toCompletableFuture().join();
            ensure(boundActorId.equals(boundAuth.actorId()), "SM-D6 bound auth actor mismatch");

            shadow.connect().submit().toCompletableFuture().join();
            Contracts.ActorAuthRes shadowAuth = shadow
                .request(new Contracts.ActorAuthReq(shadowActorId, shadowProfile))
                .submit(Contracts.ActorAuthRes.class).toCompletableFuture().join();
            ensure(shadowActorId.equals(shadowAuth.actorId()), "SM-D6 shadow auth actor mismatch");

            var shadowPush = shadow.expectNone(Contracts.ActorPushNotify.class)
                .within(Duration.ofMillis(400))
                .submit();
            var boundPush = bound.waitFor(Contracts.ActorPushNotify.class)
                .submit(Contracts.ActorPushNotify.class);
            Contracts.ActorEchoRes reply = bound
                .request(new Contracts.ActorEchoReq("push-bound-only", 20, boundProfile))
                .metadata("actor-id", boundActorId)
                .submit(Contracts.ActorEchoRes.class).toCompletableFuture().join();
            Contracts.ActorPushNotify notify = boundPush.toCompletableFuture().join().payload();

            ensure(boundActorId.equals(reply.actorId()), "SM-D6 reply actor mismatch");
            ensure(boundActorId.equals(notify.actorId()), "SM-D6 push actor mismatch");
            ensure("push:push-bound-only".equals(notify.value()), "SM-D6 push value mismatch");
            shadowPush.toCompletableFuture().join();

            waitForPlayAEvidence(List.of("ActorEntryReq|play-a|entry|" + boundActorId + "/push-bound-only"));
            System.out.println("scenario SM-D6 passed");
        } catch (Exception error) {
            throw new IllegalStateException("bound session push isolation scenario failed", error);
        } finally {
            closeQuietly(bound);
            closeQuietly(shadow);
        }

    }
}

File: framework/languages/java/e2e-kotlin/SpotService/Client/src/main/kotlin/systems/zlink/e2e/kotlin/spotservice/client/scenarios/SmD6Scenario.kt
package systems.zlink.e2e.kotlin.spotservice.client.scenarios

import systems.zlink.framework.kotlin.*

import java.time.Duration
import kotlinx.coroutines.CoroutineStart
import kotlinx.coroutines.async
import kotlinx.coroutines.supervisorScope
import systems.zlink.e2e.kotlin.spotservice.Contracts
import systems.zlink.e2e.kotlin.spotservice.Env
import systems.zlink.e2e.kotlin.spotservice.client.support.createStreamConnector
import systems.zlink.e2e.kotlin.spotservice.client.support.ensure

internal object SmD6Scenario {
    suspend fun run() {
        val bound = createStreamConnector(Env.get("e2e.stream.a.endpoint"))
        val shadow = createStreamConnector(Env.get("e2e.stream.b.endpoint"))
        try {
            val boundProfile = Contracts.ActorProfile("Bound", 6, listOf("bound"))
            val shadowProfile = Contracts.ActorProfile("Shadow", 6, listOf("shadow"))
            bound.connect().await()
            bound
                .request(Contracts.ActorAuthReq("actor-sm-d6", boundProfile))
                .awaitReply<Contracts.ActorAuthRes>()
            shadow.connect().await()
            shadow
                .request(Contracts.ActorAuthReq("actor-sm-d6-shadow", shadowProfile))
                .awaitReply<Contracts.ActorAuthRes>()

            val (reply, notify) = supervisorScope {
                val shadowPush = async(start = CoroutineStart.UNDISPATCHED) {
                    shadow.expectNone<Contracts.ActorPushNotify>("ActorPushNotify")
                        .within(Duration.ofMillis(400))
                        .await()
                }
                val boundPush = async(start = CoroutineStart.UNDISPATCHED) {
                    bound.waitFor<Contracts.ActorPushNotify>().await()
                }
                val reply = bound
                    .request(Contracts.ActorEchoReq("push-bound-only", 20, boundProfile))
                    .awaitReply<Contracts.ActorEchoRes>()
                val notify = boundPush.await().payload()
                shadowPush.await()
                reply to notify
            }

            ensure(reply.actorId == "actor-sm-d6", "SM-D6 reply actor mismatch")
            ensure(notify.actorId == "actor-sm-d6", "SM-D6 push actor mismatch")
            ensure(notify.value == "push:push-bound-only", "SM-D6 push value mismatch")
            println("scenario SM-D6 passed")
        } finally {
            try {
                bound.close().await()
            } catch (_: Exception) {
            }
            try {
                shadow.close().await()
            } catch (_: Exception) {
            }
        }
    }
}
